log4j2/src/main/java/com/google/common/flogger/backend/log4j2/Log4j2LogDataFormatter.java

```diff
@@ -16,25 +16,33 @@
 
 package com.google.common.flogger.backend.log4j2;
 
-import static java.util.logging.Level.WARNING;
-
+import com.google.common.flogger.LogContext;
+import com.google.common.flogger.backend.BaseMessageFormatter;
 import com.google.common.flogger.backend.LogData;
 import com.google.common.flogger.backend.MessageUtils;
 import com.google.common.flogger.backend.Metadata;
-import com.google.common.flogger.backend.SimpleMessageFormatter;
-import com.google.common.flogger.backend.SimpleMessageFormatter.SimpleLogHandler;
+
 import java.util.logging.Level;
 
-/** Helper to format LogData */
+import static com.google.common.flogger.backend.log4j2.Log4j2LogMessageFormatter.*;
+import static java.util.logging.Level.WARNING;
+
+/**
+ * Helper to format LogData
+ */
 final class Log4j2LogDataFormatter {
-  private Log4j2LogDataFormatter() {}
+  private Log4j2LogDataFormatter() {
+  }
 
   /**
    * Formats the log message and any metadata for the given {@link LogData}, calling the supplied
    * receiver object with the results.
    */
   static void format(LogData logData, SimpleLogHandler receiver) {
-    SimpleMessageFormatter.format(logData, receiver);
+    Metadata metadata = logData.getMetadata();
+    Throwable thrown = metadata.findValue(LogContext.Key.LOG_CAUSE);
+    StringBuilder out = BaseMessageFormatter.appendFormattedMessage(logData, new StringBuilder());
+    receiver.handleFormattedLogMessage(logData.getLevel(), out.toString(), thrown);
```

Comment:
Actually, I wanted to avoid changing too much unless I really have to or a code owner indicates it is ok to refactor, which seems to be the case. Hence I did not evaluate how much value it gives (I need to admit).

> Personally I don't think this method, or the receiver API needs to exist. Same with the formatBadLogData() method (I would just move the appendLogData() into the Log4j2LogMessageFormatter class). Actually if that's not a method available in MessageUtils, I might just add it there to avoid the duplication (can do that as a follow up though).

I noticed that in https://github.com/google/flogger/pull/272 somebody already worked on refactoring Log4j2LogMessageFormatter class. Maybe it makes sense to wait until this PR is merged and then update our branch accordingly.